### PR TITLE
Add opensearch promotion workflow and jenkins libs for tar

### LINKED ISSUE #1332: Take a release candidate build and publish its artifacts onto artifacts.opensearch.org
We need a system to accept a url and then pick up the RC build and promote those artifacts to production, they might need to be signed for this to complete.

Needs to work on tarball + native plugins

### PR DESCRIPTION
**Pull request by peterzhuamazon**
Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

### Description
Add opensearch promotion workflow and jenkins libs for tar
 
### Issues Resolved
#1332
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (gaiksaya) [inline jenkins/promotion/tar-promotion.jenkinsfile:15]: Let's not keep any default values, let it fail rather than copying some wrong files and pushing to prod

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (gaiksaya) [inline jenkins/promotion/tar-promotion.jenkinsfile:21]: Same as above

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (gaiksaya) [inline jenkins/promotion/tar-promotion.jenkinsfile:51]: Maybe add success failure to post prod promotion messages into slack channel. Will be useful when someone runs it by mistake and we know about it off release cycle

File: vars/promoteArtifacts.groovy
Comment (gaiksaya) [inline vars/promoteArtifacts.groovy:18]: Are we planning to run this differently for each platform/architecture? Like twice, once for x64 and once for arm64?

File: vars/promoteArtifacts.groovy
Comment (gaiksaya) [inline vars/promoteArtifacts.groovy:36]: Upload location needs to be corrected before merge (I think you are still testing).

File: vars/promoteArtifacts.groovy
Comment (gaiksaya) [inline vars/promoteArtifacts.groovy:47]: Same as above

**Review by gaiksaya — CHANGES_REQUESTED** [review]
Overall this looks great. Will save us manual work especially after signing part (cc: @abhinavGupta16)
Thanks!

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile:51]: This will not be run automatically or trigger by anything.
As this is the last step before release out to public, it requires user to 2PR and watching the progress.
I can add alerts to slack but that can be a improvement later on.

Thanks.

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:36]: The Jenkins workflow is currently running through my fork.
The correction will be done soon.

Thanks.

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:47]: Will change.

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile]: fixed.

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile]: fixed.

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:18]: yes, and that is ok because you will have time that you dont have both arch for a platform.
e.g. no arm64 for windows release only x64.

**Comment by dblock** [comment]
Add a test for the new function in vars and this looks good to me.

File: vars/promoteArtifacts.groovy
Comment (peternied) [inline vars/promoteArtifacts.groovy:20]:
```suggestion
    withAWS(role: "${ARTIFACT_DOWNLOAD_ROLE_NAME}", roleAccount: "${AWS_CENTRAL_ARTIFACT_ACCOUNT}", duration: 900, roleSessionName: 'jenkins-session') {
```

File: vars/promoteArtifacts.groovy
Comment (peternied) [inline vars/promoteArtifacts.groovy:36]: nit, move these parameters on to there own line

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peternied) [inline jenkins/promotion/tar-promotion.jenkinsfile:22]: For follow up, I think we can vastly reduce the number of parameters if we feed in a build manifest file as it has many of these parameters in it, and its a better reference.

**Review by peternied — COMMENTED** [review]
This is great for the current release - can you please create an issue on the outstanding items so we can fast follow

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:20]: In our Jenkins it is mistakenly named `AWS_ACCOUNT_PUBLIC` and we need a greater efforts to change all of our Jenkins file if we want to change that name now.

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:36]: Added to issue.

File: vars/promoteArtifacts.groovy
Comment (peterzhuamazon) [reply vars/promoteArtifacts.groovy:48]: Added to issue.

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile:32]: yes, and that is ok because you will have situation that you dont have both arch for a platform.
e.g. no arm64 for windows release only x64.

Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile:32]: We can definitely add another trigger just like `check-sha` jenkinsfile to multi-trigger it.

Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile:32]: Added to issue.

File: jenkins/promotion/tar-promotion.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/promotion/tar-promotion.jenkinsfile:22]: Yes, as this is a startup for the task, it is a bit complicated in the logic.
Can fine-tune later will add to the issue.
Thanks.

**Review by gaiksaya — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
@dblock @peternied @gaiksaya here is the improvement issue for this task: https://github.com/opensearch-project/opensearch-build/issues/1355

Please review again so we can unblock this and merge.

Thanks.

**Review by peternied — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Pull request merged**
